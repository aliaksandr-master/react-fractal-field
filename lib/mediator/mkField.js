export const mkMeta = () => ({
  valid: true,
  touched: false,

  active: false,
  activated: false,

  submitted: false,
  submitting: false,
  submitFailed: false,
  submitSuccess: false,
  submittedTimes: 0,

  error: null,
  errorsIn: [],
  hasException: false,
  hasFormatException: false,
  hasNormalizeException: false,
  hasValidateException: false
});

export default ({
  id,
  complexValueType = null,
  parentID = null,
  autoClean = false,
  shareMeta = true,
  shareValue = true,
  value = undefined,
  name = null
}) => ({
  // permanent
  id,
  parentID,
  autoClean,
  shareMeta,
  shareValue,
  complexValueType, // if not null component might have named children

  // free transform
  name,

  value,

  meta: mkMeta(),

  children: [],
  version: 0,
  metaVersion: 0,
  valueVersion: 0
});
