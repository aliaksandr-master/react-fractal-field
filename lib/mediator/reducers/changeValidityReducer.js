import isEqual from 'lodash/isEqual';
import recalcValidity from '../recalcValidity';
import recalcValue from '../recalcValue';



export default () => (state, { fieldID, payload: error }) => {
  const field = state.fields[fieldID];

  if (isEqual(field.meta.error, error)) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        meta: {
          ...field.meta,
          error
        }
      }
    }
  };

  const newState = recalcValidity(state, field.id);

  if (newState === state) {
    return state;
  }

  return recalcValue(newState, field.parentID, field.id);
};
