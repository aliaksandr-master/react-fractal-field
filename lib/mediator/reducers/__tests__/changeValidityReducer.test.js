/*eslint-env jest*/
import changeValidityReducer from '../changeValidityReducer';
import mkField from '../../mkField';


test('changeValidityReducer', () => {
  expect(changeValidityReducer()({}, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();

  expect(changeValidityReducer()({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123, payload: 'some' })).toMatchSnapshot();
});
