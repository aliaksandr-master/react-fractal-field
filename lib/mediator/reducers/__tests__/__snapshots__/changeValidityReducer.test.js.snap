// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeValidityReducer 1`] = `Object {}`;

exports[`changeValidityReducer 2`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": null,
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": "some",
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": false,
    },
    "metaVersion": 2,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": undefined,
    "valueVersion": 0,
    "version": 2,
  },
}
`;
